File: 329-longest-increasing-path-in-a-matrix/longest-increasing-path-in-a-matrix.cpp
class Solution {
public:
    int n, m;
    vector<vector<int>> dp;
    vector<pair<int, int>> directions = {{-1, 0}, { 1, 0 }, {0, -1}, {0, 1}};
    bool isValidCell(int row, int col)
    {
        return row >= 0 && row < n && col >= 0 && col < m;
    }

    int dfs(int row, int col, vector<vector<int>>& matrix)
    {
        if(!isValidCell(row, col))
        return 0;

        if(dp[row][col] != -1)
        return dp[row][col];

        // path at least includes current cell
        int maxLen = 1; 
        for (auto [dx, dy] : directions) 
        {
            int nr = row + dx;
            int nc = col + dy;
            if(isValidCell(nr, nc) && matrix[nr][nc] > matrix[row][col])
            maxLen = max(maxLen, 1 + dfs(nr, nc, matrix));
        }

        return dp[row][col] = maxLen;
    }

    int longestIncreasingPath(vector<vector<int>>& matrix) 
    {
        n = matrix.size();
        m = matrix[0].size();
        dp.resize(n, vector<int>(m, -1));

        int maxLength = 0;
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < m; j++)
            maxLength = max(maxLength, dfs(i, j, matrix));
        }
        return maxLength;
    }
};

File: 329-longest-increasing-path-in-a-matrix/longest-increasing-path-in-a-matrix.java
class Solution {
    private final int[][] directions = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}};
    private int n, m;

    private boolean isValidCell(int row, int col) {
        return row >= 0 && row < n && col >= 0 && col < m;
    }

    private int solve(int[][] matrix) {
        int[][] outdegree = new int[n][m];
        Queue<int[]> queue = new LinkedList<>();

        // Step 1: compute outdegree for each cell
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                for (var direction : directions) {
                    int x = i + direction[0];
                    int y = j + direction[1];
                    if (isValidCell(x, y) && matrix[x][y] > matrix[i][j]) {
                        outdegree[i][j]++;
                    }
                }
        // Step 2: push cells with outdegree == 0 (local maxima)
                if (outdegree[i][j] == 0) {
                    queue.offer(new int[]{i, j});
                }
            }
        }



        // Step 3: BFS layer by layer
        int longestIncreasingPath = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int s = 0; s < size; s++) {
                int[] node = queue.poll();
                int row = node[0], col = node[1];

                for (var direction : directions) {
                    int x = row + direction[0];
                    int y = col + direction[1];
                    if (isValidCell(x, y) && matrix[x][y] < matrix[row][col]) {
                        outdegree[x][y]--;
                        if (outdegree[x][y] == 0) {
                            queue.offer(new int[]{x, y});
                        }
                    }
                }
            }
            longestIncreasingPath++;
        }
        return longestIncreasingPath;
    }

    public int longestIncreasingPath(int[][] matrix) {
        n = matrix.length;
        m = matrix[0].length;
        return solve(matrix);
    }
}
